src/limit.c

```diff
@@ -0,0 +1,80 @@
+#include "erlinit.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+
+static rlim_t str_to_limit(const char *s)
+{
+  if(strlen(s) == 0)
+    return RLIM_INFINITY; // if the string is empty, set use infinity
+  else if(strcmp(s, "unlimited"))
+    return RLIM_INFINITY; // unlimited is a special string used by prlimit(1)
+  else
+    return atoi(s);
+}
+
+static int str_to_resource(const char *s)
+{
+  if(strcmp(s, "core") == 0)
+    return RLIMIT_CORE;
+  else if (strcmp(s, "data") == 0)
+    return RLIMIT_DATA;
+  else if (strcmp(s, "nice") == 0)
+    return RLIMIT_NICE;
+  else if (strcmp(s, "fsize") == 0)
+    return RLIMIT_FSIZE;
+  else if (strcmp(s, "sigpending") == 0)
+    return RLIMIT_SIGPENDING;
+  else if (strcmp(s, "memlock") == 0)
+    return RLIMIT_MEMLOCK;
+  else if (strcmp(s, "rss") == 0)
+    return RLIMIT_RSS;
+  else if (strcmp(s, "nofile") == 0)
+    return RLIMIT_NOFILE;
+  else if (strcmp(s, "msgqueue") == 0)
+    return RLIMIT_MSGQUEUE;
+  else if (strcmp(s, "rtprio") == 0)
+    return RLIMIT_RTPRIO;
+  else if (strcmp(s, "stack") == 0)
+    return RLIMIT_STACK;
+  else if (strcmp(s, "cpu") == 0)
+    return RLIMIT_CPU;
+  else if (strcmp(s, "nproc") == 0)
+    return RLIMIT_NPROC;
+  else if (strcmp(s, "as") == 0)
+    return RLIMIT_AS;
+  else if (strcmp(s, "locks") == 0)
+    return RLIMIT_LOCKS;
+  else if (strcmp(s, "rttime") == 0)
+    return RLIMIT_RTTIME;
+  else
+    warn("Unrecognized resource %s", s);
+
+  return -1;
+}
+
+void create_limits()
+{
+  char *temp = options.limits;
+  debug("%s", temp);
+  while(temp) {
+      const char *resource_name = strsep(&temp, ":");
+      const char *soft = strsep(&temp, ":");
+      const char *hard = strsep(&temp, ":");
+      const char *data = strsep(&temp, ";");
+      if(resource_name && soft && hard) {
+        int resource = str_to_resource(resource_name);
+        struct rlimit limit;
+        limit.rlim_cur = str_to_limit(soft);
+        limit.rlim_max = str_to_limit(hard);
+        if(prlimit(getpid(), resource, &limit, NULL) != 0)
```

Comment:
Any reason not to use `setrlimit`? According to the man page, it sounds the same as this call to `prlimit`. If `prlimit` is important, the man page also says that you can specify `0` to refer to the current process.